### PR TITLE
Add in-process concurrency support to the PS worker

### PR DESCRIPTION
**Pull request by daxian-dbw**
Add in-process concurrency support to the PS worker:
- Use `PSThreadOptions.ReuseThread` for the `InitialSessionState` when creating `Runspace`, so that every `PowerShellManager` only creates one thread and then reuse it afterwards. The default behavior is to create a new thread every time `PowerShell.Invoke` is called.
- Update `RequestProcessor` to process `InvocationRequest` in asynchronously via tasks.
- Implement `PowerShellManagerPool` using `AsyncConcurrentPool`
   - make upper bound of the pool configurable via an environment variable
   - make the pool able to expand in a lazy way
   - synchronize the call to `CheckoutIdleWorker` when there is no idle `PowerShellManager` available, in a non-blocking way.
- Update the `RpcLogger` and make every `PowerShellManager` have its own logger instance.
   - also update the way to set the `RequestId` and `InvocationId` for logger. The original way to setup the context only works for single-thread design.

### CODE REVIEW

File: src/PowerShell/PowerShellManagerPool.cs [hunk 1]
```diff
@@ -15,37 +20,80 @@ namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
...
+        private int _poolSize;
 
         /// <summary>
         /// Constructor of the pool.
         /// </summary>
-        internal PowerShellManagerPool(ILogger logger)
+        internal PowerShellManagerPool(MessagingStream msgStream)
         {
-            _logger = logger;
+            string upperBound = Environment.GetEnvironmentVariable("InProcConcurrencyUpperBound");
```
Comment (TylerLeonhardt) [inline src/PowerShell/PowerShellManagerPool.cs]: Should we maybe use something that includes the word "Worker" or "PowerShell" just so people know where this is actually being used?

File: src/PowerShell/PowerShellManagerPool.cs [hunk 2]
```diff
@@ -15,37 +20,80 @@ namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
...
+            if (string.IsNullOrEmpty(upperBound) || !int.TryParse(upperBound, out _upperBound))
```
Comment (TylerLeonhardt) [inline src/PowerShell/PowerShellManagerPool.cs]: Maybe worth logging that we couldn't parse the upperBound if int.TryParse fails. Warning level.

It also might be worth logging that "upperBound was not set - using the default". Verbose/Trace level.

File: src/PowerShell/PowerShellManagerPool.cs [hunk 3]
```diff
@@ -15,37 +20,80 @@ namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
...
+
+            try
+            {
+                logger.SetContext(requestId, invocationId: null);
+                _pool.Add(new PowerShellManager(logger));
+                _poolSize = 1;
+            }
+            finally
+            {
+                logger.ResetContext();
```
Comment (TylerLeonhardt) [inline src/PowerShell/PowerShellManagerPool.cs]: Why is the ResetContext needed here if we instantiate a new RpcLogger every time?

File: src/PowerShell/PowerShellManagerPool.cs [hunk 4]
```diff
@@ -15,37 +20,80 @@ namespace Microsoft.Azure.Functions.PowerShellWorker.PowerShell
...
+                if (_poolSize < _upperBound &&
+                    Interlocked.Increment(ref _poolSize) <= _upperBound)
+                {
+                    // If the pool hasn't reached its bounded capacity yet, then
+                    // we create a new item and return it.
+                    var logger = new RpcLogger(_msgStream);
+                    logger.SetContext(requestId, invocationId);
+                    psManager = new PowerShellManager(logger);
+                }
+                else
```
Comment (TylerLeonhardt) [inline src/PowerShell/PowerShellManagerPool.cs]: another potential log. Warning level, I think.

**Comment by daxian-dbw** [comment]
Close the PR, will submit a new one from my fork.

**Pull request closed without merging**